### PR TITLE
Ensure certgen handles already-existing secrets correctly

### LINKED ISSUE #2150: Failed to Apply contour-certgen Job Manifest!
**What steps did you take and what happened:**

`kubectl apply -f https://raw.githubusercontent.com/projectcontour/contour/v1.1.0/examples/contour/02-job-certgen.yaml`

This error appears:
```
The Job "contour-certgen" is invalid: spec.template: Invalid value: core.PodTemplateSpec{ObjectMeta:v1.ObjectMeta{Name:"", GenerateName:"", Namespace:"", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"app":"contour-certgen", "controller-uid":"24fb4489-a03c-4a34-baae-a4e7b017954d", "job-name":"contour-certgen"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:core.PodSpec{Volumes:[]core.Volume(nil), InitContainers:[]core.Container(nil), Containers:[]core.Container{core.Container{Name:"contour", Image:"docker.io/projectcontour/contour:latest", Command:[]string{"contour", "certgen", "--incluster", "--kube"}, Args:[]string(nil), WorkingDir:"", Ports:[]core.ContainerPort(nil), EnvFrom:[]core.EnvFromSource(nil), Env:[]core.EnvVar{core.EnvVar{Name:"CONTOUR_NAMESPACE", Value:"", ValueFrom:(*core.EnvVarSource)(0xc00238d600)}}, Resources:core.ResourceRequirements{Limits:core.ResourceList(nil), Requests:core.ResourceList(nil)}, VolumeMounts:[]core.VolumeMount(nil), VolumeDevices:[]core.VolumeDevice(nil), LivenessProbe:(*core.Probe)(nil), ReadinessProbe:(*core.Probe)(nil), StartupProbe:(*core.Probe)(nil), Lifecycle:(*core.Lifecycle)(nil), TerminationMessagePath:"/dev/termination-log", TerminationMessagePolicy:"File", ImagePullPolicy:"Always", SecurityContext:(*core.SecurityContext)(nil), Stdin:false, StdinOnce:false, TTY:false}}, EphemeralContainers:[]core.EphemeralContainer(nil), RestartPolicy:"Never", TerminationGracePeriodSeconds:(*int64)(0xc00b1d6300), ActiveDeadlineSeconds:(*int64)(nil), DNSPolicy:"ClusterFirst", NodeSelector:map[string]string(nil), ServiceAccountName:"contour-certgen", AutomountServiceAccountToken:(*bool)(nil), NodeName:"", SecurityContext:(*core.PodSecurityContext)(0xc0001e1810), ImagePullSecrets:[]core.LocalObjectReference(nil), Hostname:"", Subdomain:"", Affinity:(*core.Affinity)(nil), SchedulerName:"default-scheduler", Tolerations:[]core.Toleration(nil), HostAliases:[]core.HostAlias(nil), PriorityClassName:"", Priority:(*int32)(nil), PreemptionPolicy:(*core.PreemptionPolicy)(nil), DNSConfig:(*core.PodDNSConfig)(nil), ReadinessGates:[]core.PodReadinessGate(nil), RuntimeClassName:(*string)(nil), Overhead:core.ResourceList(nil), EnableServiceLinks:(*bool)(nil), TopologySpreadConstraints:[]core.TopologySpreadConstraint(nil)}}: field is immutable
```

**What did you expect to happen:**

No error.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

I was attempting to upgrade from Contour v1.0.1 to v1.1.0, and this error happened.

**Environment:**

- Contour version: 1.1.0
- Kubernetes version: (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-13T11:23:11Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-13T11:13:49Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```
- Kubernetes installer & version:
```
kubeadm version: &version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-13T11:20:25Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: UpCloud 2 core 4 GB RAM
- OS (e.g. from `/etc/os-release`):
```
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
VERSION_CODENAME=stretch
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

### PR DESCRIPTION
**Pull request by youngnick**
Fixes #2150 by ensuring that the `contour certgen` Job will succeed if the secrets already exist.

Signed-off-by: Nick Young <ynick@vmware.com>

### CODE REVIEW

File: internal/certgen/certgen.go [hunk 1]
```diff
@@ -103,6 +104,10 @@ func writeCACertKube(client *kubernetes.Clientset, namespace string, cert []byte
 	secret := newCertOnlySecret("cacert", namespace, "cacert.pem", cert)
 	_, err := client.CoreV1().Secrets(namespace).Create(secret)
 	if err != nil {
+		if k8serrors.IsAlreadyExists(err) {
```
Comment (jpeach) [inline internal/certgen/certgen.go]: Nit: would this style be more canonical?
```Go
if _, err := client.CoreV1().Secrets(namespace).Create(secret); if err != nil {
        ...
}
```

**Comment by jpeach** [comment]
@youngnick Is there a place to document this?

File: internal/certgen/certgen.go (hunk 1 above)
Comment (youngnick) [reply internal/certgen/certgen.go]: Won't hurt. I'll change it.

**Comment by youngnick** [comment]
It's what we always said the behaviour was, and the inspiration behind having certgen being a Job in the first place. I've obviously missed a big test case when doing this originally, 🤦‍♂.

I guess my answer is, "I'm not sure".

**Comment by jpeach** [comment]
> On 30 Jan 2020, at 2:02 pm, Nick Young <notifications@github.com> wrote:
> 
> I guess my answer is, "I'm not sure".

Then should we file an issue to document certgen?

**Comment by youngnick** [comment]
#2183 created. I'll merge this now and work on that.

**Comment by ryanelian** [comment]
Does certgen recreates existing certificates if expired?

**Comment by jpeach** [comment]
> On Jan 31, 2020, at 4:28 PM, Ryan Elian <notifications@github.com> wrote:
> 
> Does certgen recreates existing certificates if expired?

No. To regenerate the certificates you would need to delete the secrets first.

**Pull request merged**